### PR TITLE
Pods 8442 allow access

### PR DESCRIPTION
**Pull request by PavelVjalicin**

### CODE REVIEW

File: test/controllers/actions/AllowAccessActionSpec.scala [hunk 1]
```diff
@@ -304,5 +307,37 @@ class AllowAccessActionSpec
...
+          .test(OptionalDataRequest(fakeRequest, "id", Some(UserAnswers(Json.obj())), psaId, pspId))
+          .map { case Some(value) => false
+          case None => true
+          }
+      }
+      def test(withSrn: Boolean) = {
+        hasAccess(allowPsa = true, allowPsp = false, withSrn)(isPsa = true).map { hasAccess => assert(hasAccess, true)}
+        hasAccess(allowPsa = true, allowPsp = true, withSrn)(isPsa = true).map { hasAccess => assert(hasAccess, true)}
+        hasAccess(allowPsa = false, allowPsp = true, withSrn)(isPsa = true).map { hasAccess => assert(hasAccess, false)}
+        hasAccess(allowPsa = true, allowPsp = false, withSrn)(isPsa = false).map { hasAccess => assert(hasAccess, false)}
```
Comment (darbyo) [inline test/controllers/actions/AllowAccessActionSpec.scala:334]: This isnt exhaustive of the actual possibilities, in order to support this implementation and how it relies and specific combinations in the boolean match can we add the other scenarios. To give a guard until the refactor happens

**Review by darbyo — CHANGES_REQUESTED** [review]

File: app/controllers/actions/AllowAccessAction.scala [hunk 2]
```diff
@@ -60,28 +62,65 @@ abstract class AllowAccessAction(srn: Option[String],
...
+        case (Some(true), _) => true
+        case (_, Some(true)) => true
+        case _ => false
       }
     }
-  }.getOrElse(errorHandler.onClientError(request, NOT_FOUND, "").map(Some.apply))
+
+    accessAllowed.flatMap {
+      case true => Future.successful(None)
+      case false => errorHandler.onClientError(request, NOT_FOUND, "").map(Some.apply)
```
[the remark this answers is not in the document]
Comment (darbyo) [reply app/controllers/actions/AllowAccessAction.scala:110]: same as we added retrospectively to the other services

File: app/connectors/PensionsSchemeConnector.scala [hunk 3]
```diff
@@ -84,11 +84,11 @@ class PensionsSchemeConnectorImpl @Inject()(http: HttpClient, config: FrontendAp
     }
   }
 
-  def checkForAssociation(psaId: String, srn: String)
+  def checkForAssociation(userId: String, srn: String, isPsa: Boolean)
                          (implicit headerCarrier: HeaderCarrier,
                           ec: ExecutionContext, request: RequestHeader): Future[Either[HttpResponse, Boolean]] = {
     val headers: Seq[(String, String)] =
-      Seq(("psaId", psaId), ("schemeReferenceNumber", srn), ("Content-Type", "application/json"))
+      Seq((if(isPsa) "psaId" else "pspId", userId), ("schemeReferenceNumber", srn), ("Content-Type", "application/json"))
```
[the remark this answers is not in the document]
Comment (PavelVjalicin) [reply app/connectors/PensionsSchemeConnector.scala:91]: Done.

File: app/controllers/actions/AllowAccessAction.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (PavelVjalicin) [reply app/controllers/actions/AllowAccessAction.scala]: Done

File: test/controllers/actions/AllowAccessActionSpec.scala (hunk 1 above)
Comment (PavelVjalicin) [reply test/controllers/actions/AllowAccessActionSpec.scala:334]: Not sure what is meant by this.
Should I fully simulate a controller with all of the action methods?

Comment (PavelVjalicin) [reply test/controllers/actions/AllowAccessActionSpec.scala:334]: Introduced false false tests.

**Review by darbyo — APPROVED** [review]

**Review by jamdade-sharad — APPROVED** [review]
pspOnPageLoad() method-call for psp users then I think allowPsa=false is needed, but if already tested and functionally working fine with allowPsa=true then fine.

File: app/controllers/CheckYourAnswersMembersController.scala [hunk 4]
```diff
@@ -53,7 +53,7 @@ class CheckYourAnswersMembersController @Inject()(override val messagesApi: Mess
     }
 
   def pspOnPageLoad(srn: String): Action[AnyContent] =
```
[the remark this answers is not in the document]
Comment (PavelVjalicin) [reply app/controllers/CheckYourAnswersMembersController.scala:55]: Thanks Sharad.

[the remark this answers is not in the document]
Comment (PavelVjalicin) [reply app/controllers/CheckYourAnswersMembersController.scala:55]: We decided not to change the above.

**Pull request merged**